lib/lightning/accounts/user.ex

```diff
@@ -164,14 +164,74 @@ defmodule Lightning.Accounts.User do
     |> put_change(:role, :superuser)
   end
 
-  def validate_email(changeset) do
+  @doc """
+  Validates that a user has access to all projects associated with a credential.
+  Returns a changeset with any project access errors.
+  """
+  def validate_project_access(changeset, credential) do
+    case get_field(changeset, :email) do
+      nil ->
+        changeset
+
+      email ->
+        case Lightning.Repo.exists?(User |> where(email: ^email)) do
+          false ->
+            changeset
+
+          true ->
+            user = Lightning.Accounts.get_user_by_email(email)
+
+            projects =
+              credential
+              |> Lightning.Credentials.diff_project_credentials_and_project_users(
+                user
+              )
+              |> Lightning.Projects.get_projects_by_ids()
+
+            if Enum.empty?(projects) do
+              changeset
+            else
+              projects_names = Enum.map_join(projects, ", ", & &1.name)
+
+              add_error(
+                changeset,
+                :email,
+                "User doesn't have access to these projects: #{projects_names}"
+              )
+            end
+        end
+    end
+  end
+
+  def validate_not_same_user(changeset, field, current_user, opts \\ []) do
+    message = opts[:message] || "Cannot be the same as current user"
+    value = get_field(changeset, field)
+    current_value = Map.get(current_user, field)
+
+    cond do
+      is_nil(value) ->
+        changeset
+
+      value == current_value ->
+        add_error(changeset, field, message)
+
+      true ->
+        changeset
+    end
```

Comment:
the `current_user` will always have an email so two conditions for `value == current_value` true or false would be enough. As this function is used to check equality based on the email, would you mind starting with a specific one (without the `field` parameter)? If need in the future in could be generalised unless you know that it will be or that it is very likely to be used with another field. Abstract functions are more flexible but specific ones have a more clear purpose which makes easier to maintain.

Reply:
Thanks, this is also changed and I agree that we do not need to generalize too early.